# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Asmaa-eg/flyrank-internship-ml/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I’m choosing the Refresh / Content Opportunity lane. This feels like the best fit because the starter data contains observable signals about content age, visibility, and trend, and the goal is to help a team decide which pages are worth reviewing first. I’m especially interested in whether a simple ranking approach can surface pages that look stale but still have enough demand to matter.

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import os
import subprocess
import sys
from pathlib import Path

import pandas as pd

REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"


def find_csv():
    roots = [Path.cwd(), *Path.cwd().parents]
    roots += [Path("/content"), Path("/content/" + REPO_DIR), Path("/mnt/data")]
    for base in roots:
        candidate = base / "data" / "raw" / "content_refresh_anonymized.csv"
        if candidate.exists():
            return candidate
    return None


csv_path = find_csv()
if csv_path is None:
    if "google.colab" in sys.modules:
        if not os.path.isdir(REPO_DIR):
            subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
        os.chdir(REPO_DIR)
        csv_path = find_csv()

if csv_path is None:
    raise FileNotFoundError("Could not find data/raw/content_refresh_anonymized.csv. Please make sure the repo is available in this environment.")


df = pd.read_csv(csv_path)
print("Rows:", len(df))
print("Declining rate:", round((df["trend_direction"].str.lower() == "down").mean(), 3))
print("Pages with impressions > 0:", int((df["impressions_90d"] > 0).sum()))
print("Average impressions:", round(df["impressions_90d"].mean(), 2))

Rows: 30000
Declining rate: 0.542
Pages with impressions > 0: 30000
Average impressions: 5200.37


## 2. The question: decision, action, cost of a wrong call

The decision this work could improve is which pages should be reviewed first. The action would be taken by a content or SEO team that has limited time and wants to focus on pages that are both visible and at risk of decline. A wrong recommendation would waste time on pages that do not need attention, which means less efficient review effort and slower recovery on the pages that really matter.

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Pages with impressions >= 500:", int((df["impressions_90d"] >= 500).sum()))
print("Pages with age >= 180 days:", int((df["content_age_days"] >= 180).sum()))
print("Pages with both conditions:", int(((df["impressions_90d"] >= 500) & (df["content_age_days"] >= 180)).sum()))


Pages with impressions >= 500: 16726
Pages with age >= 180 days: 17986
Pages with both conditions: 9929


## 3. Quick look at the data (2-3 real numbers)
The starter data looks relevant for this lane because the sample is large, the decline signal is common, and there is clear visibility and age variation across pages. Two real numbers that stand out are: the dataset contains 30,000 pages, and 54.2% of pages are labeled as declining. The average impressions value is 5,200.37, which suggests there is meaningful exposure across the sample.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Declining rate:", round((df["trend_direction"].str.lower() == "down").mean(), 3))
print("Average impressions:", round(df["impressions_90d"].mean(), 2))
print("Median content age (days):", round(df["content_age_days"].median(), 2))

Declining rate: 0.542
Average impressions: 5200.37
Median content age (days): 236.0


## 4. Careful words: what I can and can't claim

I can support an observed, directional claim about which pages appear more likely to need review based on signals such as age, visibility, and trend. I cannot claim causal proof or say that I can predict Google’s behavior with certainty. This work is meant to support decision-making, not to prove a universal rule.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print("Observed / directional claim:")
print("Pages that are older and still visible appear to be promising review candidates, but this is a decision-support signal rather than causal proof.")

Observed / directional claim:
Pages that are older and still visible appear to be promising review candidates, but this is a decision-support signal rather than causal proof.


## Self-check

Before you submit, confirm each line honestly:

- [.] Every section above is filled — markdown thinking AND the code that backs it
- [.] The notebook runs top to bottom with no errors (Runtime → Run all)
- [.] No client names, URLs, or private queries anywhere
- [.] My claims use careful words: observed, measured, directional, decision-support
- [. ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.